crates/orchestrator/src/network_spec/relaychain.rs

```diff
@@ -0,0 +1,118 @@
+use configuration::{
+    shared::{
+        resources::Resources,
+        types::{Arg, AssetLocation, Chain, Command, Image},
+    },
+    RelaychainConfig,
+};
+
+use super::node::NodeSpec;
+use crate::{
+    errors::OrchestratorError,
+    generators::chain_spec::{ChainSpec, Context},
+    shared::types::ChainDefaultContext,
+};
+
+/// A relaychain configuration spec
+#[derive(Debug, Clone)]
+pub struct RelaychainSpec {
+    /// Chain to use (e.g. rococo-local).
+    pub(crate) chain: Chain,
+
+    /// Default command to run the node. Can be overriden on each node.
+    pub(crate) default_command: Option<Command>,
+
+    /// Default image to use (only podman/k8s). Can be overriden on each node.
+    pub(crate) default_image: Option<Image>,
+
+    /// Default resources. Can be overriden on each node.
+    pub(crate) default_resources: Option<Resources>,
+
+    /// Default database snapshot. Can be overriden on each node.
+    pub(crate) default_db_snapshot: Option<AssetLocation>,
+
+    /// Default arguments to use in nodes. Can be overriden on each node.
+    pub(crate) default_args: Vec<Arg>,
+
+    // chain_spec_path: Option<AssetLocation>,
+    pub(crate) chain_spec: ChainSpec,
+
+    /// Set the count of nominators to generator (used with PoS networks).
+    pub(crate) random_nominators_count: u32,
+
+    /// Set the max nominators value (used with PoS networks).
+    pub(crate) max_nominations: u8,
+
+    /// Nodes to run.
+    pub(crate) nodes: Vec<NodeSpec>,
+}
+
+impl RelaychainSpec {
+    pub fn from_config(config: &RelaychainConfig) -> Result<RelaychainSpec, OrchestratorError> {
+        // Relaychain main command to use, in order:
+        // set as `default_command` or
+        // use the command of the first node.
+        // If non of those is set, return an error.
+        let main_cmd = if let Some(cmd) = config.default_command() {
+            cmd
+        } else if let Some(first_node) = config.nodes().first() {
+            let Some(cmd) = first_node.command() else {
+                return Err(OrchestratorError::InvalidConfig("Relaychain, either default_command or command in the first node needs to be set.".to_string()));
+            };
+
+            cmd
+        } else {
+            return Err(OrchestratorError::InvalidConfig(
+                "Relaychain without nodes and default_command isn't set.".to_string(),
+            ));
+        };
```

Comment:
It could be a good idea to simplify this part with functionnal code, I feel like the error is duplicated, from what I understand in the above comment we could have something like this:

```suggestion
        let main_cmd = config
            .default_command() // take default command if it exists
            .or(config.nodes().first().and_then(|node| node.command())) // take the first node command if there is no default command 
            .ok_or(Err(OrchestratorError::InvalidConfig("Relaychain, either default_command or first node with a command needs to be set.".to_string())))?; // returns error if both case above didn't worked
```

wdyt ?

Reply:
Yes, make sense. I will change it as feedback.